subsys/fs/nvs/nvs.c

```diff
@@ -6,814 +6,764 @@
 {
 	int rc;
-	off_t addr;
-	struct _nvs_data_hdr head;
-	u16_t hdr_len, slt_len, sec_hdr_len;
+	size_t ate_size, empty_len;
+	struct nvs_ate ate;
+	u32_t addr;
 
-	rc = 0;
 
 	k_mutex_lock(&fs->nvs_lock, K_FOREVER);
 
-	/* fill previous sector with data to jump to the next sector */
-	head.id = NVS_ID_SECTOR_END;
-	hdr_len = _nvs_len_in_flash(fs, sizeof(struct _nvs_data_hdr));
-	slt_len = _nvs_len_in_flash(fs, sizeof(struct _nvs_data_slt));
-	sec_hdr_len = _nvs_len_in_flash(fs, sizeof(struct _nvs_sector_hdr));
-	head.len = fs->sector_size
-		   - (fs->write_location & (fs->sector_size - 1))
-		   - slt_len - hdr_len + sec_hdr_len;
-	rc = nvs_flash_write(fs, fs->write_location, &head, sizeof(head));
-	if (rc) {
-		goto out;
-	}
-
-	/* advance to next sector for writing */
-	addr = (fs->write_location & ~(fs->sector_size - 1));
-	_nvs_addr_advance(fs, &addr, fs->sector_size);
+	ate_size = _nvs_al_size(fs, sizeof(struct nvs_ate));
 
-	/* initialize the new sector */
-	rc = _nvs_sector_init(fs, addr);
-	if (rc) {
-		goto out;
-	}
-
-	/* data copy */
-	addr = (fs->write_location & ~(fs->sector_size - 1));
-	_nvs_addr_advance(fs, &addr, fs->sector_size);
-	/* Do we need to advance the entry_sector, if so we need to copy */
-	if ((addr & ~(fs->sector_size - 1)) ==
-		fs->entry_sector * fs->sector_size) {
-		SYS_LOG_DBG("Starting data copy...");
-		_nvs_entry_sector_advance(fs);
-		rc = _nvs_gc(fs, addr);
+	/* step through the sectors to find the last sector */
+	for (u16_t i = 0; i < fs->sector_count; i++) {
+		addr = (i << 16) + fs->sector_size - ate_size;
+		fs->ate_wra = addr;
+		rc = _nvs_prev_ate(fs, &addr, &ate);
 		if (rc) {
-			SYS_LOG_DBG("Quit data copy - gc error");
-			goto out;
+			goto end;
 		}
-		rc = _nvs_flash_erase(fs, addr, fs->sector_size);
-		if (rc) {
-			SYS_LOG_DBG("Quit data copy - flash erase error");
-			goto out;
+		rc = _nvs_flash_cmp_const(fs, addr - ate_size, 0x00, ate_size);
+		if (!rc) {
+			/* we are just before (or after) sector end */
+			continue;
+		}
+		rc = _nvs_flash_cmp_const(fs, addr - ate_size, 0xff, ate_size);
+		if (!rc) {
+			break;
 		}
-		SYS_LOG_DBG("Done data copy - no error");
 	}
-	rc = 0;
 
-out:
-	k_mutex_unlock(&fs->nvs_lock);
-	return rc;
-}
+	fs->ate_wra = addr;
+	fs->data_wra = addr & 0xffff0000;
 
-int nvs_compute_crc(struct nvs_fs *fs, const struct nvs_entry *entry,
-		    u16_t *crc16)
-{
-	int rc;
-
-	*crc16 = 0xFFFF;
-	for (int i = 0; i < entry->len; i += fs->write_block_size) {
-		u8_t buf[fs->write_block_size];
-
-		rc = nvs_flash_read(fs, entry->data_addr + i, buf, sizeof(buf));
+	/* Normally addr points to a location that does not contain all ff, so
+	 * ate_size needs to be substracted to get ate_wra. However there is an
+	 * exveption, when addr is the very end of a sector; then we keep
+	 * fs->ate_wra.
+	 */
+	rc = _nvs_flash_cmp_const(fs, addr, 0xff, ate_size);
+	if (rc < 0) {
+		goto end;
+	}
+	if (rc) {
+		/* fs->ate_wra not at sector end */
+		fs->ate_wra -= ate_size;
```

Comment:
shouldn't nvs iterates more for seek ate_wra? this assume that ate_wra referencing the second ate in a sector.

Reply:
no ate_wra is found by _nvs_prev_ate(), more iterations are not needed. ate_wra is not referencing the second ate in a sector.

Reply:
ok